docs/rfcs/20220119_raft_replication_simplification.md

```diff
@@ -0,0 +1,162 @@
+`append_entries_requests` to the follower in parallel with leader disk flush.
+In this case raft implementation may have multiple requests in flight per follower. 
+(currently number of requests pending is limited by per follower semaphore). 
+
+Leader ack consistency replication is implemented differently 
+control is returned to the caller immediately after data are written 
+to the underlying `storage::log`. On the next heartbeat follower 
+is recognized to be behind the leader and is being recovered by `raft::recovery_stm`. 
+The `recovery_stm` reads a chunk of data and sends it to the follower 
+in a loop (single request at a time).
+
+## Purpose of the simplification
+
+There are multiple reasons that makes the simplification necessary and demanded, 
+the most important are:
+
+- code simplification
+- unification of approaches
+- better handling of requests with mixed consistency level (current they cause latency spikes)
+
+Also some future features that will benefit from simplification are
+
+- in memory acks
+- leader being remote to data 
+
+
+## Proposed solution
+
+Proposed solution assumes that both currently supported consistency levels will 
+be handled in the same way with use of `replicate_batcher` and `replicate_entries` stm. 
+This approach has several advantages over the one currently 
+used for `leader_ack` consistency level. 
+
+Current implementation of `quorum_acks` replication is robust and pretty 
+well tested. No large code changes are required to use the 
+`replicate_batcher` and `replicate_stm` to handle `leader_acks` replication. 
+The way that the `replicate_stm` handles `append_entries` requests dispatching 
+and leader log writes will allow to make them parallel in future. 
+
+Additionally using `replicate_stm` prevents us from creating the 
+append entries request twice since batches coming from `replicate_batcher` 
+are not first handed of to disk and then read again but rather shared to 
+generate leader disk write and follower `append_entries` requests. 
+This approach significantly reduce the `storage::batch_cache` pressure and may 
+also reduce the disk usage (in some cluster with limited memory cache hit 
+ratio is not always 1.0 even if there are no reads other than reads issued by 
+`replicate_stm`).
+
+### Backpressure
+
+In current `quorum_ack` replication implementation the backpressure is 
+propagated from the followers to the `raft::replicate` caller 
+(Kafka Produce request handler effectively). Currently the backpressure 
+propagation is based on the per follower semaphore since RPC layer 
+lacks backpressure propagation mechanism. 
+When `replicate_entries` stm is unable to acquire follower dispatch units 
+it will wait holding `replicate_batcher` memory reservation lock so that 
+it can not accept writes indefinitely. When `replicate_batcher` memory 
+reservation semaphore is exhausted it prevents first stage of `replicate_stages` 
+from finishing not allowing for more requests to be processed. 
```

Comment:
```suggestion
In the current `quorum_ack` replication implementation, the backpressure is 
propagated from the followers to the `raft::replicate` caller 
(the Kafka Produce request handler effectively). Currently, the backpressure 
propagation is based on the per-follower semaphore since the RPC layer 
lacks a backpressure propagation mechanism. 
When `replicate_entries` stm is unable to acquire follower dispatch units, 
it will wait, holding the `replicate_batcher` memory reservation lock so that 
it can not accept writes indefinitely. When the `replicate_batcher` memory 
reservation semaphore is exhausted, it prevents the first stage of `replicate_stages` 
from finishing, not allowing more requests to be processed. 
```